src/mesh/coordinates.cxx

```diff
@@ -978,6 +978,21 @@ void Coordinates::outputVars(Datafile& file) {
   getParallelTransform().outputVars(file);
 }
 
+const Field2D& Coordinates::zlength() const {
+  BOUT_OMP(critical)
+  if (not zlength_cache) {
+    zlength_cache = std::make_unique<Field2D>(0., localmesh);
+
+#if BOUT_USE_METRIC_3D
+    BOUT_FOR_SERIAL(i, dz.getRegion("RGN_ALL")) { (*zlength_cache)[i] += dz[i]; }
+#else
+    (*zlength_cache) = dz * nz;
+#endif
+  }
```

Comment:
```suggestion
  if (not zlength_cache) {
    BOUT_OMP(critical)
    if (not zlength_cache) {
      auto tmp = std::make_unique<Field2D>(0., localmesh);

#if BOUT_USE_METRIC_3D
    BOUT_FOR_SERIAL(i, dz.getRegion("RGN_ALL")) { (*tmp)[i] += dz[i]; }
#else
     (*tmp) = dz * nz;
#endif
      zlength_cache = tmp;
    }
  }
```

I haven't tested it, but I think something like that should avoid the locking in most cases - as the final update is (mostly?) atomic ...
However, not sure it is worth it - as it should be easy enough to pull out of OMP loops ...

Reply:
Could `Delp2` be changed to use the tridag interface which takes `kwave` directly?

Reply:
`tmp` is needed so `zlength_cache` is immediately valid, and not first set to `0`.

Reply:
@dschwoerer Ah, I see, ok

@d7919 In principle, but currently the `kwave` overload is `protected` in `Laplacian` and not public.

Looking further, `Delp2` doesn't use OpenMP, so this wouldn't affect it anyway.

Reply:
I'm of the opinion this change probably isn't worth it, as all of the potentially affected library code is now fixed or not using OpenMP, and I suspect most/all user code that uses `zlength` won't be using OpenMP